### PR TITLE
Added pre-login XSRF protection to the XSRF plugin

### LINKED ISSUE #130: Add pre-login XSRF protection
The XSRF plugin will need to protect login forms from attacks as this can also introduce vulnerabilities See [this link](https://stackoverflow.com/questions/6412813/do-login-forms-need-tokens-against-csrf-attacks) for more context.

### PR DESCRIPTION
**Pull request by maramihali**
Fixes #130, #136 

The XSRF protection can now also protect login forms by setting a custom cookie to distinguish between users. The cookie will then be used as an identifier in the verification and generation algorithm.

### CODE REVIEW

**Review by mattiasgrenfeldt — APPROVED** [review]
One nit :+1:

File: safehttp/plugins/xsrf/xsrf.go [hunk 1]
```diff
@@ -63,24 +56,52 @@ func Token(r *safehttp.IncomingRequest) (string, error) {
...
 // In case of state changing requests (all except GET, HEAD and OPTIONS), it
 // checks for the presence of an XSRF token in the request and validates it
-// based on the user ID associated with the request.
+// based on the custom cookie.
 //
 // For authorized requests, it adds a cryptographically safe XSRF token to the
-// incoming request. It can be later extracted using Token.
+// incoming request. This can be later extracted using Token.
 func (i *Interceptor) Before(w *safehttp.ResponseWriter, r *safehttp.IncomingRequest, cfg interface{}) safehttp.Result {
-	userID, err := i.Identifier.UserID(r)
```
Comment (kele) [inline safehttp/plugins/xsrf/xsrf.go:77]: I don't get this part. Why is this removed?

Comment (mihalimara22) [reply safehttp/plugins/xsrf/xsrf.go:77]: So, the previous plugin version was vulnerable to pre-login CSRF. An attack would go as follows: I am logged into your bank account and the attacker also has an actual account on the banking website. Then, the attacker can forge a login request in my browser and trick me into using their account. Then, they'll have access to stuff like my transaction history while I am logged in into their account.
 
This would be allowed by the previous plugin because the call to `Identifier.UserID(r)` request would return the ID corresponding to the attacker account even if I am the one navigating and the token that is injected in the form has also been generated for the attacker's account.

Therefore, we need a way to identify that I am the one navigating the attacker's account. One way to do this is: the first time I navigate the banking website with a state preserving request, a noonce-based cookie is generated with SameSite Strict. Then, any xsrf token validation fails because the token was generated with the attacker's cookie and we validate with our cookie.

Hopefully this made sense (@empijei please correct me if I misunderstood something).

Comment (kele) [reply safehttp/plugins/xsrf/xsrf.go:77]: But now you've taken away the feature of extracting a user ID from a request, possibly via some lookup operations.

Comment (mihalimara22) [reply safehttp/plugins/xsrf/xsrf.go:77]: AFAIU that is not needed anymore.

Comment (kele) [reply safehttp/plugins/xsrf/xsrf.go:77]: Discussed this with @empijei offline.

Long story short:
- you're right, we don't really need the `UserIdentifier` - it wasn't clear to me why
- please document why need both the cookies + token injection in the response

File: safehttp/plugins/xsrf/xsrf.go [hunk 2]
```diff
@@ -63,24 +56,52 @@ func Token(r *safehttp.IncomingRequest) (string, error) {
...
+	needsValidation := !statePreservingMethods[r.Method()]
+	c, err := r.Cookie(tokenCookie)
```
Comment (kele) [inline safehttp/plugins/xsrf/xsrf.go:91]: The "tokenCookie" doesn't really tell match what it is. In reality, it is some kind of a user ID / session ID. Could you try to find a better name?

File: safehttp/plugins/xsrf/xsrf.go [hunk 3]
```diff
@@ -63,24 +56,52 @@ func Token(r *safehttp.IncomingRequest) (string, error) {
 	return tok.(string), nil
 }
 
+func addTokenCookie(w *safehttp.ResponseWriter) (*safehttp.Cookie, error) {
+	buf := make([]byte, 20)
+	_, err := rand.Read(buf)
+	if err != nil {
+		panic(fmt.Errorf("failed to generate entropy using crypto/rand/RandReader: %v", err))
```
Comment (kele) [inline safehttp/plugins/xsrf/xsrf.go:63]:
```suggestion
		panic(fmt.Errorf("crypto/rand.Read: %v", err))
```

File: safehttp/plugins/xsrf/xsrf.go [hunk 4]
```diff
@@ -63,24 +56,52 @@ func Token(r *safehttp.IncomingRequest) (string, error) {
...
+		if err != nil {
+			// An error is returned when the plugin fails to Set the Set-Cookie
+			// header in the response writer as this is a server misconfiguration.
+			return w.ServerError(safehttp.StatusInternalServerError)
+		}
 	}
 
-	actionID := r.Method() + " " + r.URL.Path()
-	needsValidation := !statePreservingMethods[r.Method()]
+	actionID := r.URL.Path()
```
Comment (kele) [inline safehttp/plugins/xsrf/xsrf.go:104]: Why this change?

File: safehttp/plugins/xsrf/xsrf.go (hunk 2 above)
Comment (kele) [inline safehttp/plugins/xsrf/xsrf.go:91]: Consider a longer name than "c", it is used over ~30 lines.

**Review by kele — CHANGES_REQUESTED** [review]

File: safehttp/plugins/xsrf/xsrf.go (hunk 4 above)
Comment (maramihali) [reply safehttp/plugins/xsrf/xsrf.go:104]: Having the method there is error-prone since we could visit the same path using different HTTP methods. I discussed offline with @empijei who said that it's fine if the `actionID` is just the path.

File: safehttp/plugins/xsrf/xsrf.go [hunk 5]
```diff
@@ -63,24 +56,55 @@ func Token(r *safehttp.IncomingRequest) (string, error) {
 	return tok.(string), nil
 }
 
+func addCookieID(w *safehttp.ResponseWriter) (*safehttp.Cookie, error) {
+	buf := make([]byte, 20)
+	_, err := rand.Read(buf)
+	if err != nil {
+		panic(fmt.Errorf("crypto/rand.Read: %v", err))
```
Comment (bsiegert) [inline safehttp/plugins/xsrf/xsrf.go:63]: I am nervous about this `panic` in the request path. Can this return the error instead?

**Review by bsiegert — APPROVED** [review]

File: safehttp/plugins/xsrf/xsrf.go (hunk 5 above)
Comment (maramihali) [reply safehttp/plugins/xsrf/xsrf.go]: I changed it to return an error instead, but shouldn't a system failure result in a panic?

**Review by empijei — APPROVED** [review]

**Review by kele — APPROVED** [review]
Thank you. :)

**Comment by kele** [comment]
Please rebase and we can merge.

**Comment by maramihali** [comment]
Rebased, ready for merge.
